src/dialogsformaction.cpp

```diff
@@ -87,12 +87,12 @@ bool DialogsFormAction::process_operation(Operation op,
 {
 	switch (op) {
 	case OP_OPEN: {
-		const unsigned int pos = dialogs_list.get_position();
+		const unsigned int pos = dialogs_list.get_movement_control().get_position();
```

Comment:
I think it's ugly that the users of `ListWidget` have to know about `ListMovementControl`. How about inheriting `ListWidget` from `ListMovementControl`, so these methods become directly accessible? Then `ListMovementControl`'s interface can be split into public and protected parts, which would better explain how the class is supposed to be used. This shouldn't affect testability as it's still easy to pass a stub of `IList` in tests. What do you think?

Reply:
Sounds good.
However, I just tried to execute this idea and ran into some issues.

The `ListMovementControl` requires the `IList` reference directly in its constructor (if we want to store it as a reference in a field).
That gives some issues in the `ListWidget` if it derives from the `ListMovementControl`.
Deriving it requires that the `ListMovementControl` is initialized first, but it already needs access to the `IList` (`ListWidgetBackend`) which is a field in the later constructed `ListWidget`.
Other related options I thought about had similar issues (or created a lot of ugliness in the designs of the classes)

I did find a way to get rid of all the `get_movement_control()` calls by providing dereferencing operators (`operator->` and `operator*`.
I think it looks quite natural in the places where it is used.
Do you find this solution acceptable?

Reply:
Oh, my bad, I should have realized that the initialization order would mess that up!

If the call to `get_movement_control()` was *too* explicit to my taste, then the call to `operator->` is too implicit :) It looks quite mysterious and unexpected that I have to dereference the object in order to get to some of its methods.

I have two more ideas though. If you aren't interested, then please roll back to `get_movement_control()` and continue with that.

The first idea is to make `ListMovementControl` a mixin: replace `IList& list` member with `protected: virtual IList& get_list() = 0`, then inherit `ListWidget` from it and implement the method as `ILIst& get_list() override { return backend; }`.

The second idea requires more changes, but actually results in less code than any of the previous ones:

1. make `ListMovementControl` a template, parameterized by the backend type. Inherit `ListMovementControl` from the backend type, both to get to its methods and "re-export" its public interface;
2. `ListWidget` becomes `using ListWidget = ListMovementControl<ListWidgetBackend>;` 
3. backend methods that are only relevant to movement control can become `protected` now;
4. drop `IList` because it has no purpose anymore.

In essence, we build up `ListWidget`'s interface from the ground up: `ListWidgetBackend` gives us the two `stfl_replace_*` methods and `get_width()`, and then `ListMovementControl` gives us the movement methods.

What do you think about those?

Reply:
No problem, always nice trying out new things :)
I just pushed an attempt at implementing your second idea.
Is this what you had in mind?

I might need to take another look at the current tests to see if they can be cleaned up a bit (for now, I just stopped as soon as I got them to compile again)
However, I probably won't have time until next weekend.

Reply:
Yep, looks precisely as I envisioned it, great job!